classlib: add number of decimal places to SimpleNumber -asTimeString

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->
After discussion in #4661, I'd like to propose adding an argument to control number of decimal places for ```aNumber.AsTimeString```.

Before I write tests for this, I wanted to see if others think this is a worthwhile addition.

## Types of changes

<!-- Delete lines that don't apply -->

- (Documentation)
- New feature

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] All tests are passing
  - [x] new tests were added and passing
- [x] Updated documentation
- [x] This PR is ready for review
- [x] cleanup commit history

brianlheim commented:
so i think this is a good idea overall. i'm a little concerned about adding too many features to this function, but date/time conversion functions tend to be that way, i think by necessity. and there's no reason milliseconds should be the default.

the inverse function `asSecs` should also be updated to accept strings in this format.

dyfer commented:
> the inverse function `asSecs` should also be updated to accept strings in this format.

`asSecs` already accepts any number of floating point digits.

brianlheim commented:
> asSecs already accepts any number of floating point digits.

good! then the only thing to update is the documentation :)

> Return a Float based on converting a time string in format (ddd):hh:mm:ss.sss

dyfer commented:
> > Return a Float based on converting a time string in format (ddd):hh:mm:ss.sss

an example already has this: `"-23:12.3456".asSecs;`; how would you phrase the explanation to indicate variable number of digits? or would adding more examples be the way to go?

brianlheim commented:
yes i saw that example, but it's in conflict with the wording in the documentation. up to you how to phrase it, but i would say "hh:mm:ss.z", where z can be any sequence of decimal digits. is the fractional part optional? in that case i would notate it as "hh:mm:ss(.z)". also worth checking what happens with "hh:mm:ss." (decimal point but no fractional part) and documenting that too.

dyfer commented:
@brianlheim I think I addressed your concerns, could you take a look?
(`"hh:mm:ss."` - string with trailing period - is not explicitly documented, but showed in examples)

I also realize that this method breaks with scientific notation. I don't see a way to fix it ATM. Do you think that should be documented, fixed (how?) or is it fine as is, since it's rather an edge case?
```supercollider
9e-05.asTimeString(precision: 0.00001, decimalPlaces: 5)
// -> 00:00:00.9e-05
```

brianlheim commented:
> I also realized that this method breaks with scientific notation. I don't see a way to fix it ATM. Do you think that should be documented, fixed (how?) or is it fine as is, since it's rather an edge case?


that's a total non-starter. there's got to be a way to do this, it's such a basic formatting task. i just checked and even `Float.asStringPrec` doesn't do it. if we had #3570 we could accomplish this very easily. for now you could just fix it inline, with a note saying that when we solve #3570  (i.e., add sprintf) we can use that directly?

brianlheim commented:
> fixed (how?)

https://www.geeksforgeeks.org/convert-floating-point-number-string/

except we only need the code to convert the fractional part

dyfer commented:
@brianlheim thanks for the feedback and the example code. Just to clarify, my thinking behind this possibly not be a task for this PR was mainly because, as you noted, SC currently doesn't have means of controlling the format of strings obtained from floats. I'll update the PR shortly.

brianlheim commented:
> Just to clarify, my thinking behind this possibly not be a task for this PR was mainly because, as you noted, SC currently doesn't have means of controlling the format of strings obtained from floats.

i understand that point, but we also should only knowingly be shipping incomplete/broken features as a last resort. my order of preferences would be:

1. fix #3570 and then implement this feature
2. implement this feature with a noted workaround for #3570
3. don't implement this feature
4. implement this feature in a broken state

which is, in abstract form:

1. add working feature with no technical debt
2. add working feature with technical debt
3. don't add feature
4. add broken feature

i would order these differently depending on the amounts of technical debt and brokenness involved. my view is, some features we are ok with implementing in an imperfect state, if they maybe require more time or feedback to get right but add a lot of value right away. similarly, we can ship things that add technical debt as long as the route to working it down is clear and easy. here, the workaround is pretty obvious, and the way to iron it out later is also obvious, so i'm backing option (2).

dyfer commented:
@brianlheim sounds good! Just to be clear, I wasn't attacking your point, just explaining mine.

Does the implementation and the comment look fine for you now?

brianlheim commented:
there was no offense taken, i am a little tired so i'm sorry if that came off as defensive.

it looks good at a glance, i would appreciate if you could add some tests so we can be sure of the behavior.

brianlheim commented on HelpSource/Classes/SimpleNumber.schelp:
```diff
@@ -368,16 +368,16 @@ method:: asAudioRateInput
 argument:: precision
-accuracy of the millisecond format; the string will always be formatted with 3 decimal places for
-milliseconds. Minimum value: 0.001.
+accuracy of the fraction of seconds; since the number of decimal places is also an argument, minimum value of code::precision:: is code::10.pow(decimalPlaces.neg)::, i.e. code::0.001:: for code::decimalPlaces = 3::
```
this kinda makes it sound like values under that minimum are not allowed -- could you reword it to clarify that those values will be replaced with the minimum (or clamped to it)?

brianlheim commented on HelpSource/Classes/SimpleNumber.schelp:
```diff
@@ -368,16 +368,16 @@ method:: asAudioRateInput
 method:: asTimeString
-Produces a time string in the format code::ddd:hh:mm:ss.sss::, interpreting the receiver as time in
-seconds. See link::Classes/String#-asSecs:: for the inverse function.
+Produces a time string in the clock format inspired by ISO 8601 time interval display (truncated representation) code::ddd:hh:mm:ss.sss::, interpreting the receiver as time in seconds. See link::Classes/String#-asSecs:: for the inverse function.
```
maybe good to note here that the exact format is configurable, and not fixed at `ddd:hh:mm:ss.sss`

brianlheim commented on HelpSource/Classes/SimpleNumber.schelp:
```diff
@@ -368,16 +368,16 @@ method:: asAudioRateInput
 argument::dropDaysIfPossible
 a link::Classes/Boolean::. If set to code::true::, and the number of days in the formatted string
-would be 0, that section is instead ommitted and the string is formatted as code::hh:mm:ss.sss::.
+would be 0, that section is instead ommitted and the string is formatted as code::hh:mm:ss.sss::
```
since we're editing this -- can you fix the typo? ommitted->omitted. and maybe just leave off the "is formatted as hh:mm:ss.sss" since that isn't technically true anymore.

brianlheim commented on HelpSource/Classes/SimpleNumber.schelp:
```diff
@@ -368,16 +368,16 @@ method:: asAudioRateInput
-would be 0, that section is instead ommitted and the string is formatted as code::hh:mm:ss.sss::.
+would be 0, that section is instead ommitted and the string is formatted as code::hh:mm:ss.sss::
+argument::decimalPlaces
+number of decimal places representing fraction of seconds; if code::0:: (minimum value), the string is formatted as code::(ddd:)hh:mm:ss::
```
probably good to note what happens if you pass in something < 0.

brianlheim commented on HelpSource/Classes/String.schelp:
```diff
@@ -248,18 +248,21 @@ code::
 
 method::asSecs
-Return a link::Classes/Float:: based on converting a time string in format (ddd):hh:mm:ss.sss. This is the inverse method to link::Classes/SimpleNumber#-asTimeString::.
+Return a link::Classes/Float:: based on converting a time string in format (ddd:)hh:mm:ss(.z), where z is any number of digits. This is the inverse method to link::Classes/SimpleNumber#-asTimeString::.
```
"sequence of digits" would be more accurate, i think. also, can you wrap the format string in `code:: ::` to match the documentation for `asTimeString`?

brianlheim commented on HelpSource/Classes/String.schelp:
```diff
@@ -248,18 +248,21 @@ code::
-(45296.asTimeString).asSecs;
+"00:00:59.900".asSecs; //hh:mm:ss.zzz
+"1:1:1.1".asSecs; //h:m:s.z
+"001:00:00:00.001".asSecs; //ddd:hh:mm:ss.zzz
```
please add spaces after the `//`s

brianlheim commented on HelpSource/Classes/String.schelp:
```diff
@@ -248,18 +248,21 @@ code::
+"-23:12.346".asSecs;
+"-1:00:00:00".asSecs; // neg with days
+"12:-34:56".asSecs; // warns
+(45296.789.asTimeString).asSecs; //inverse of aNumber.asTimeString
```
add space after `//`

brianlheim commented on testsuite/classlibrary/TestSimpleNumber.sc:
```diff
@@ -99,7 +106,28 @@ TestSimpleNumber : UnitTest {
+	test_asTimeString_smallNumberScientificNotation {
+		var expected = "00:00:00.0000010";
+		var totalTime = 1e-06;
+		var actual = totalTime.asTimeString(precision: 0.0000001, decimalPlaces: 7);
```
for me anything with more zeroes than `0.001` is difficult to read  -- i would personally prefer `1e-7` here but it's not something you need to change here

brianlheim replied:
oh, sorry for the miscommunication -- i was actually talking about the float in:

```supercollider
totalTime.asTimeString(precision: 0.0000001, decimalPlaces: 7);
                                  // ^ how many zeroes?
```

i was fine with the `totalTime = 1e-6` part :)

brianlheim commented on SCClassLibrary/Common/Math/SimpleNumber.sc:
```diff
@@ -705,13 +711,13 @@ SimpleNumber : Number {
+			mseconds = "." ++ (number.frac.round(precision) * pow(10, decimalPlaces)).asInteger.asString.padLeft(decimalPlaces, "0");
+		}, {
+			mseconds = "";
+		});
```
i think i left a note about this on another PR, generally we write this with trailing closure syntax. also, you can assign the result of `if` to an object, so i think this would be more clearly written as:

```supercollider
mseconds = if(decimalPlaces > 0) { "." ++ ... } { "" };
```

brianlheim commented on SCClassLibrary/Common/Math/SimpleNumber.sc:
```diff
@@ -705,13 +711,13 @@ SimpleNumber : Number {
+		seconds = (decimal % 60).asString.padLeft(2, "0");
+		if(decimalPlaces > 0, {
+			// this could be simplified once sclang gains sprintf functionality, see issue #3570
+			mseconds = "." ++ (number.frac.round(precision) * pow(10, decimalPlaces)).asInteger.asString.padLeft(decimalPlaces, "0");
```
this will result in bad output if you pass in `decimalPlaces` > 11 or 12, because you'll overflow a 32-bit integer:

```supercollider
0.1.asTimeString(precision: 0.001, decimalPlaces: 13)
// -> 00:00:00.00-2147483648
```

a safer way to do it is accumulate a digit at a time while multiplying the fractional part by 10, although i believe you may lose precision that way. i'm not very familiar with numerical analysis, so i can't say for certain what the best way to approach it would be. perhaps it's not very important here anyway. but this function shouldn't produce garbage on valid inputs.

redFrik replied:
here's something similar i've been using and that doesn't overflow...
```
(
~fixDec= {|val, numDecimals= 2|  //float to string with fixed number of decimals
        var str= val.round(0.1**numDecimals).asString;
        var num= str.size-str.indexOf($.)-1;
        str.extend(str.size+numDecimals-num, $0);
};
)

a= 0.1.rand
~fixDec.value(a, 13)
```

brianlheim replied:
thanks @redFrik ! but it fails on inputs that result in scientific notation: `~fixDec.value(1e-5, 13`

dyfer commented:
@brianlheim I think I addressed all your comments and fixed the bug with large values of `decimalPlaces`. Thank you for finding that! I even added a test for it :)

brianlheim commented on SCClassLibrary/Common/Math/SimpleNumber.sc:
```diff
@@ -705,13 +711,24 @@ SimpleNumber : Number {
+				curFrac = curFrac - curInt;
+				curDecimalPlaces = curDecimalPlaces - 1;
+			};
+			str;
```
i think this would probably make sense as a separate private function. otherwise, the vars declared inside the `if` will cause warnings from the compiler. also makes it easier to read the function generally.

dyfer replied:
warning from sclang?

separate function should also be hidden from help; happy to do it if that's still preferred for a "temporary" (before `sprintf`) solution

brianlheim replied:
> warning from sclang?

yes. see https://doc.sccode.org/Classes/LanguageConfig.html#*postInlineWarnings

> separate function should also be hidden from help

yep, it should be private and still in this class

brianlheim commented on SCClassLibrary/Common/Math/SimpleNumber.sc:
```diff
@@ -705,13 +711,24 @@ SimpleNumber : Number {
+				curInt = curFrac.asInteger;
+				str = str ++ curInt.asString;
+				curFrac = curFrac - curInt;
+				curDecimalPlaces = curDecimalPlaces - 1;
```
why do you round to `curPrec` every pass? that seems unnecessary on the face of it

dyfer replied:
This is to compensate for lost precision/rounding error when "cutting" the `.frac` at every pass. Without this at least one test failed.

brianlheim commented:
re: precision @dyfer 

like i said, i don't know much about numerical computation / numerical analysis, but i can at least test to see whether an implementation is correct. i've been using this code to test it:

```supercollider
(
n = 6;
~inequalCount = 0;
~rounds = 100000;
postln("real\t\t\t\tasString\tasTimeString");
~rounds.do {
	
	~randVal = 1.0.rand.round(pow(10, n.neg));
	while { ~randVal.asString.contains($e) } {
		~randVal = 1.0.rand
		// weed out ones that get represented as exponential notation
	};
	~timeString = ~randVal.asTimeString(decimalPlaces: n, precision: pow(10, n.neg));
	~string = ~randVal.asString.keep(n + 2).padRight(2 + n, "0"); // 2= num chars for leading "0."
	~timeStringSuffix = ~timeString.keep(n.neg - 2); // 2= num chars for leading "0."
	
	if (~string != ~timeStringSuffix) {
		// pad to 18 to keep columns consistent
		postf("%\t%\t% (%)\n", ~randVal.asString.padRight(18, " "), ~string, ~timeStringSuffix, ~timeString);
		~inequalCount = ~inequalCount + 1;
	}
};

~ratio = ~inequalCount / ~rounds;
~ratio
)
```

for n=5 there are already a few small errors (click on 'details' to expand):

<details>
```
real				asString	asTimeString
0.33837509155273  	0.33837	0.33838 (00:00:00.33838)
0.2927485704422   	0.29274	0.29275 (00:00:00.29275)
0.40283739566803  	0.40283	0.40284 (00:00:00.40284)
0.32503616809845  	0.32503	0.32504 (00:00:00.32504)
0.10154747962952  	0.10154	0.10155 (00:00:00.10155)
```
</details>

for n=6 there are also errors where extra digits are appended, it looks like at some points `curInt` is rounded up to 10:

<details>
```
real				asString	asTimeString
0.78              	0.780000	.7710000 (00:00:00.7710000)
0.48              	0.480000	.4710000 (00:00:00.4710000)
0.06              	0.060000	.0510000 (00:00:00.0510000)
0.38191080093384  	0.381910	0.381911 (00:00:00.381911)
0.29619550704956  	0.296195	0.296196 (00:00:00.296196)
0.0063178539276123	0.006317	0.006318 (00:00:00.006318)
0.13              	0.130000	.1210000 (00:00:00.1210000)
0.14              	0.140000	.1310000 (00:00:00.1310000)
```
</details>

for n=12 there are too many to print, and the `~ratio` at the end is ~0.9999.

the first kind of errors are not so bad to me, but the second (printing 0.771 instead of 0.78) are very very bad.

i looked at an actual implementation of `printf` and it's horrendously complex: just the floating-point portion of glibc's is over 1000 lines long! (https://github.com/lattera/glibc/blob/master/stdio-common/printf_fp.c#L205)

unless you see a way around it, i'm going to consider this blocked on #3570. it's not a heavily demanded feature, and i don't want to ship broken code.

dyfer commented:
@brianlheim thanks for a thorough test!

I think I arrived at a reasonable solution. I'm using something like `number.frac.asString` and then checking for scientific notation and treating the resulting string appropriately.

It does pass your test - although I made a small change: I think the number you compare against should also be rounded:
```supercollider
	~string = ~randVal.round(pow(10, n.neg)).asString.keep(n + 2).padRight(2 + n, "0"); // 2= num chars for leading "0."
```

Please let me know how this looks now!

dyfer commented:
> I'm thinking whether rounding should be also clamped based on the number of significant figures after floating point in the input number?

I made another change and this time I'm not using `.frac` which I believe avoids the problem of rounding the fractional part separately

brianlheim commented:
> I think the number you compare against should also be rounded:

ah! i'm doing that in the initial assignment to `~randVal`, but didn't realize i wasn't doing that inside the while loop. that probably explains some of the problems there. thank you.

i was able to find some failures with n=18:

<details>

```
real				asString	asTimeString
0.0087124109268188	0.008712410926818800	0.008712410926818900 (00:00:00.008712410926818900)
0.0078579187393188	0.007857918739318800	0.007857918739318900 (00:00:00.007857918739318900)
0.0083004236221313	0.008300423622131300	0.008300423622131400 (00:00:00.008300423622131400)
0.0089718103408813	0.008971810340881300	0.008971810340881400 (00:00:00.008971810340881400)
0.0080868005752563	0.008086800575256300	0.008086800575256400 (00:00:00.008086800575256400)
0.0086055994033813	0.008605599403381300	0.008605599403381400 (00:00:00.008605599403381400)
```

</details>

but they're all in the last significant digit which seems totally acceptable to me. it may not even be worth documenting.

as for the branches handling scientific notation.. i think something is wrong here:

```supercollider
0.000000994999.asTimeString(precision: 1e-12, decimalPlaces: 12)
-> 00:00:00.009949990000
```

at this point, i'd prefer just solving #3570 directly and using that here. this seems like a bit of a rabbit hole. but, if you want to keep working on it, you're welcome to do so. i'd just ask that you please thoroughly test your work before asking for another review.

dyfer commented:
For the record: I think the scientific notation issue is now solved. The tests for `n=18` (and larger) did not yield any errors for me, but I think I made a change that was not pushed with the last commit, sorry.

I'll revisit this once commits from  #4662 are in `develop`. Maybe by that time #3570 will be solved and could be used as a better solution. For now what I have here seems to work, but will probably need more testing.

brianlheim commented:
hey @dyfer -- i'm still interested in merging this one.  it looks like #4662 is in develop but #3570 is still unsolved. do you want to revisit this?

dyfer commented:
Thanks for looking into this @brianlheim 

I've rebased this to resolve conflicts. It seems that the current implementation passes all the cases that we've thought of before, but it would be much cleaner if we had an equivalent of `sprintf`.
